adding sample exporter

szeli1 opened the pull request:
closes #7452

This Pull request adds sample exporting to lmms that is independent from AudioEngine.

How to test:
Open lmms, load a sample into a sample clip. Open the context menu and click on the new option called "export sample buffer". A file dialog should open up. Select an audio file to export to, or create a new file by using a custom name in the file dialog. Click "open" inside the file dialog. The sample should be exported to the selected output file.

related pull request: #7557
this allows #5990 to be finished

michaelgregorius commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -376,5 +385,27 @@ bool SampleClipView::splitClip( const TimePos pos )
+	if (openFileDialog.exec() == QDialog::Accepted)
+	{
+		QStringList curSelectedFiles(openFileDialog.selectedFiles());
+		if (curSelectedFiles.isEmpty() == false)
```
```
if (!curSelectedFiles.isEmpty())
```

michaelgregorius commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -376,5 +385,27 @@ bool SampleClipView::splitClip( const TimePos pos )
+		QStringList curSelectedFiles(openFileDialog.selectedFiles());
+		if (curSelectedFiles.isEmpty() == false)
+		{
+			QString outputFileName = curSelectedFiles.first();
```
It seems that you silently drop all other files if several files are selected in the dialog. Please instead disallow the selection of multiple files in the first place.

szeli1 replied:
How to fix this?

michaelgregorius replied:
I took the liberty to push some changes with commit 3566660591f to save some back and forth. The gist of it is that `FileDialog::getSaveFileName` is used to retrieve the file name. This gives us lots of functionality automatically, e.g. the request if an existing file should be overwritten. The previous implementation did not do this which I had to learn the hard way. :sweat_smile:

It also removes the appending of the `.flac` file extension and lets the users decide the name.

michaelgregorius commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -376,5 +385,27 @@ bool SampleClipView::splitClip( const TimePos pos )
+		{
+			QString outputFileName = curSelectedFiles.first();
+			if (outputFileName.isEmpty()) { return; }
+			if (outputFileName.endsWith(".flac") == false)
```
```
if (!outputFileName.endsWith(".flac"))
```

szeli1 replied:
I would prefer using false, it isn't as hidden

michaelgregorius commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -376,5 +385,27 @@ bool SampleClipView::splitClip( const TimePos pos )
+			if (outputFileName.isEmpty()) { return; }
+			if (outputFileName.endsWith(".flac") == false)
+			{
+				outputFileName = PathUtil::stripPrefix(outputFileName) + ".flac";
```
If I understand correctly you let the users choose different file types but then save it as a Flac anyway. This is very confusing in my opinion.

szeli1 replied:
Yes, the exporter class only supports .flac and this was agreed to in discord. I can still change that, but it will add unreasonable complexity in my view.

michaelgregorius commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -376,5 +385,27 @@ bool SampleClipView::splitClip( const TimePos pos )
+				outputFileName = PathUtil::stripPrefix(outputFileName) + ".flac";
+			}
+
+			m_clip->s_sampleExporter->startExporting(outputFileName, m_clip->m_sample.buffer());
```
The view should not know any details of the clip implementation, e.g. that the clip has a member called `s_sampleExporter`. The view method also accesses the clip's sample buffer. This is a code smell called ["feature envy"](https://refactoring.guru/smells/feature-envy). Instead the `SampleClip` class should provide a method to export its buffer to a file, e.g. something like `SampleClip::exportToFile(const QString& filename)`.

The static member in `SampleClip` is likely unnecessary anyway.

szeli1 replied:
I use it since it can export in an other thread, so static is needed.

michaelgregorius commented on src/core/LmmsExporterSample.cpp:
```diff
@@ -0,0 +1,179 @@
+		outputBuffer[i + 1] = outputBuffer[i + 1] > 1.0f ? 1.0f : outputBuffer[i + 1] < -1.0f ? -1.0f : outputBuffer[i + 1];
+		i = i + channelCount;
+	}
+	size_t count = sf_writef_float(m_fileDescriptor, outputBuffer.data(), static_cast<sf_count_t>(buffer->size()));
```
IMO all the sndfile related code should be moved into a helper class that is concerned with writing LMMS buffers into files in various formats using sndfile. It could be an RAII class which
* takes a filename and the export format in its constructor and opens the file using sndfile
* provides a method to push LMMS buffers which are then written into the file
* in the destructor closes the file.

Doing so would likely also remove duplicated code in `AudioFileFlac` and `AudioFileWave`.

JohannesLorenz replied:
I would agree to Michael, but I don't understand (yet?) why you couldn't just use `AudioFileFlac` directly.

szeli1 replied:
Yes. It was decided to leave the audio engine code alone, https://github.com/LMMS/lmms/pull/5990 removed audio engine from `AudioFileFlac`, but it was closed because this is a better solution.

JohannesLorenz replied:
I guess you mean #7452 ? I think it's an interesting way to do this. Any reason why you think this is better?

szeli1 replied:
I've answered this question in discord, in short, this PR doesn't touch the audio engine, so it avoids introducing regressions. It also is more simple and robust than refactoring the Audio engine exporting code. The `AudioFileDevice` classes were designed to fit into audio engine. Also they can be done in 1 file, exporting different formats doesn't have to be separated into multiple files.

JohannesLorenz replied:
It seems @michaelgregorius has nothing to add, so I'll resolve this comment here. Feel free to reopen.

firewall1110 commented:
>It seems that https://github.com/LMMS/lmms/pull/7627 only exports the content of sample clips into files, i.e. it will not replace the recorded data with references to the exported files.

Can this be fixed with new option to **_replace_** old reference or data with reference **_to exported file_**?

michaelgregorius commented:
> > It seems that #7627 only exports the content of sample clips into files, i.e. it will not replace the recorded data with references to the exported files.
> 
> Can this be fixed with new option to **_replace_** old reference or data with reference **_to exported file_**?

We should not implement too many workflow crutches that we might never get rid of again. IMO recording to files should be implemented as fast as possible instead of some workarounds.

szeli1 commented:
> Can this be fixed with new option to replace old reference or data with reference to exported file?

Originally I couldn't do it because the exporting ran on a different thread and SampleClipView didn't wait for the exporting to be done before trying to import it.

firewall1110 commented:
> SampleClipView didn't wait for the exporting to be done

Can this **_exporting thread before quit trigger event_** connected to some SampleClip (OR ...) slot that do the same as happens when user  load file to clip?

szeli1 commented:
I will need to move the FlacExporter into a new file, but it is ready for review.

szeli1 commented:
I realized that I would rather not touch `AudioFileFlac`, because I don't fully understand how it works. For example the `writeBuffer()` function has "integer PCM encoding" built in, while `FlacExporter` doesn't and it doesn't need to have it. It can certainly be used to replace most libsndfile functions for `AudioFileFlac`, but I'm unsure if this should be done.

szeli1 commented:
@michaelgregorius It seems like you introduced a crash by removing the code that ensures ".flac" file extensions. If the name (`outputFilename`) doesn't contain an extension in the file dialog, then lmms will crash. I believe this happens because after exporting, the exported sample will be reimported by the clip, but the clip will try to import `outputFilename` without the ".flac" extension that does not exist.

michaelgregorius commented:
> @michaelgregorius It seems like you introduced a crash by removing the code that ensures ".flac" file extensions are there. If the name (`outputFilename`) doesn't contain an extension in the file dialog, then lmms will crash. I believe this happens because after exporting, the exported sample will be reimported by the clip, but the clip will try to import `outputFilename` without the ".flac" extension that does not exist.

@szeli1, I have fixed this with commit 5fa60b35503 by saving the file with the name that the users have selected and not appending the ".flac" extension.

michaelgregorius commented on include/LmmsMassExporter.h:
```diff
@@ -0,0 +1,76 @@
+
+typedef void (*callbackFn)(void*);
+
+class LmmsMassExporter
```
Some remarks about the name of the class:
* It lives in the `lmms` namespace so there is not need to prefix it with "Lmms".
* Why is it called mass exporter? To me it seems that one of its main features is that it exports data in a thread so perhaps something like `ThreadedExporter` might be more fitting.

szeli1 replied:
I like `ThreadedExportManager`, I will rename.

szeli1 commented:
> @szeli1, I have fixed this with commit https://github.com/LMMS/lmms/commit/5fa60b35503edf50742923c00e072ff77e4e2363 by saving the file with the name that the users have selected and not appending the ".flac" extension.

I don't know why do you prefer removing it, but sure, I approve if it works. I will remove the QFileInfo include.

szeli1 commented:
> @szeli1, I have fixed this with commit https://github.com/LMMS/lmms/commit/5fa60b35503edf50742923c00e072ff77e4e2363 by saving the file with the name that the users have selected and not appending the ".flac" extension.

It now works, it can export files without file extension, I don't know why is this useful, because the result will be a file without an extension, so every time the user would like to open it, they will have to select a program that can open .flac files.

CorruptVoidSoul commented:
So I tested this with a wav and an ogg file, both exported without problems and there doesn't seem to be any hearable difference.
In case it's wanted, system info is aarch64 Ubuntu running on top of Termux on a phone (samsung galaxy A25).

michaelgregorius commented:
> > @szeli1, I have fixed this with commit [5fa60b3](https://github.com/LMMS/lmms/commit/5fa60b35503edf50742923c00e072ff77e4e2363) by saving the file with the name that the users have selected and not appending the ".flac" extension.
> 
> It now works, it can export files without file extension, I don't know why is this useful, because the result will be a file without an extension, so every time the user would like to open it, they will have to select a program that can open .flac files.

The reason is that we should let users do what the want to do. The dialog already proposes the "flac" extension so this adds no extra work for the users. The default is that it is exported with the extension.

However, if the users for some reason decide that they want/need to have a file without extension then the previous implementation would make them jump through the hoops to export the file and then having to rename it manually to remove the extension. With the current implementation they simply save it without the extension and they are good.

The current implementation gives the best of both words. I guess we all have some experience with software that attempts to be clever only to do the wrong thing. So please let's not add such behavior to LMMS if it is not even necessary in some way.

michaelgregorius commented on src/core/SampleClip.cpp:
```diff
@@ -336,5 +342,19 @@ gui::ClipView * SampleClip::createView( gui::TrackView * _tv )
+{
+	if (thisObject == nullptr) { return; }
+	SampleClip* castedThis = static_cast<SampleClip*>(thisObject);
+	castedThis->setSampleFile(castedThis->m_exportedSampleName);
```
Now that I have thought a bit more about it I do not think that replacing the sample with the one we have just exported is a good idea.

Let's say you have a project with all the samples contained in the project. You now want to export the sample to a different folder outside of the project just to save it. In this case the project file would now point to a file outside of the project folder which might not be the intended effect.

IMO the export feature should simply export the file to wherever the users want. If users want to replace the sample with the exported one it should either be an option in the export dialog (initially disabled) or they should explicitly replace the samples themselves after the export.

Doing it as it is currently done will likely lead to "chaos".

JohannesLorenz replied:
I agree to this. Imagine a user saves their sample file on a mounted external disk / network folder / USB disk - then loading the song next time would fail if the device is not available anymore.

szeli1 replied:
So don't change the source location when exporting?

firewall1110 commented:
>If users want to replace the sample with the exported one it should either be an option in the export dialog (initially disabled) 

Something like this I mean when wrote:

>Can this be fixed with new option to replace old reference or data with reference to exported file?

JohannesLorenz commented:
This, again, should be merged into #5990, right?

szeli1 commented:
> This, again, should be merged into https://github.com/LMMS/lmms/pull/5990, right?

This is the plan as far as I'm aware.

JohannesLorenz commented:
> This Pull request adds sample exporting to lmms that is independent from AudioEngine.

A few questions about the scope:

1. Is sample exporting already possible with AudioEngine?
2. About using FLAC: I assume we wanted to save lossless (because REAPER saves WAV, too), and FLAC is a good compression for WAV - that was the idea?
3. The idea is to use `SampleClip::exportSampleBuffer` in #5990, correct? So, after this PR here has been reviewed, it might make sense to try it out with #5990 before we this PR to master?

szeli1 commented:
> Is sample exporting already possible with AudioEngine?

It is not possible. See #7452

> About using FLAC: I assume we wanted to save lossless (because REAPER saves WAV, too), and FLAC is a good compression for WAV - that was the idea?

Yes.

> The idea is to use SampleClip::exportSampleBuffer in https://github.com/LMMS/lmms/pull/5990, correct? So, after this PR here has been reviewed, it might make sense to try it out with https://github.com/LMMS/lmms/pull/5990 before we this PR to master?

Ultimately enabling #5990 to be finished is the end goal, I believe PR supports everything #5990 needs, but If this isn't the case, then it makes sense to merge it with #5990 instead of master

JohannesLorenz commented on src/core/SampleClip.cpp:
```diff
@@ -345,15 +345,7 @@ gui::ClipView * SampleClip::createView( gui::TrackView * _tv )
 void SampleClip::exportSampleBuffer(const QString& fileName)
 {
 	m_exportedSampleName = fileName;
```
I think the variable, and thus this function, can be removed now, to simplify the code?

szeli1 replied:
I believe this can be removed.

szeli1 replied:
This can not be removed, It is used by `SampleClipView` to export audio.

JohannesLorenz replied:
You clicked resolved but did not resolve it in code.

szeli1 replied:
I have removed `m_exportedSampleName` from the code.

JohannesLorenz commented on src/core/ThreadedExportManager.cpp:
```diff
@@ -0,0 +1,113 @@
+{
+	if (m_thread != nullptr)
+	{
+		if (m_isThreadRunning == true)
```
The `== true` might have the advantage of making pointer-to-bool conversions explicit. But here, `m_isThreadRunning` is not a pointer, but a boolean, so I think the `== true` is redundant.

szeli1 replied:
I will remove it

JohannesLorenz commented on src/core/ThreadedExportManager.cpp:
```diff
@@ -0,0 +1,113 @@
+void ThreadedExportManager::startExporting(const QString& outputLocationAndName, std::shared_ptr<const SampleBuffer> buffer, callbackFn callbackFunction, void* callbackObject)
+{
+	m_readMutex.lock();
+	m_buffers.push_back(std::make_tuple(outputLocationAndName, buffer, callbackFunction, callbackObject));
```
Also, if you use mutexes, it is usually suggested to use the RAII variant: `std::lock_guard`.

szeli1 replied:
As far as I understand, nothing is "lockless", "lockess" means avoiding being in the "lock" state for most of the time.

szeli1 replied:
Also I had problems using `LocklessRingBuffer` in the past because of it lacks documentation.

JohannesLorenz replied:
Lockless means you hit a mutex or another thing where one thread waits. The `LocklessRingBuffer` never lets any thread wait - thanks to atomic variables it always immediatelly write/reads (or fails if there is not enough space).

The documentation of `LocklessRingBuffer` is not perfect/complete, though it is just a front-end for the `ringbuffer` classes you could read the  ringbuffer.h code (Doxygen comments) or the [tests](https://github.com/JohannesLorenz/ringbuffer/tree/master/src/test).

szeli1 replied:
I have read it and it was still hard to understand, also the file follows different style which also makes it harder to read. 
Also I was incorrect about my statement, lockless structures do not force threads to wait.

Would using `std::lock_guard` be better? Currently the mutex is only locked during push back, so I would have to introduce a scope to reproduce the same behavior. I'm okay with using either one.

JohannesLorenz replied:
It's not realtime safe without ringbuffer, so you might go with a mutex (plus lock_guard), but it may cause glitches (unlikely), but it will also cause us work to fix it when we make LMMS realtime safe (or just analyze its realtime safety, see #7764).

JohannesLorenz commented on include/ThreadedExportManager.h:
```diff
@@ -0,0 +1,76 @@
+	std::vector<std::tuple<QString, std::shared_ptr<const SampleBuffer>, callbackFn, void*>> m_buffers;
+
+	volatile std::atomic<bool> m_abortExport;
+	bool m_isThreadRunning;
```
I would suggest to init it here instead of in the CTOR, same for other variables if possible.

JohannesLorenz commented on src/core/ThreadedExportManager.cpp:
```diff
@@ -0,0 +1,113 @@
+		std::tuple<QString, std::shared_ptr<const SampleBuffer>, callbackFn, void*> curBuffer = std::make_tuple(QString(""), nullptr, nullptr, nullptr);
+		thisExporter->m_readMutex.lock();
+		bool shouldExit = thisExporter->m_buffers.size() <= 0;
+		if (shouldExit == false)
```
Same here,`== false` is redundant.

szeli1 replied:
I would rather keep `== false`

JohannesLorenz replied:
It's the same as `!shouldExit`. If we write code consistently with `== false`, we could remove the `!`-operator from the C++ language.

JohannesLorenz replied:
Sorry for my previous comment, I should respect the decision of the author. I will resolve the conversation.

JohannesLorenz commented on src/core/ThreadedExportManager.cpp:
```diff
@@ -0,0 +1,113 @@
+	if (m_isThreadRunning == false)
+	{
+		stopExporting();
+		m_isThreadRunning = true;
```
Also, setting the variable here seems redundant (and possibly dangerous?) since threadedExportFunction just does that.

JohannesLorenz commented on src/core/audio/FlacExporter.cpp:
```diff
@@ -0,0 +1,110 @@
+
+FlacExporter::~FlacExporter()
+{
+	if (m_fileDescriptor == NULL) { return; }
```
We once converted all `NULL` to `nullptr` in the whole LMMS code base, so it should be converted in new PRs, too.

szeli1 replied:
I followed libsndfile's documentation, but I will replace it.

szeli1 commented:
Thanks for the review @JohannesLorenz, I will apply the suggested changes shortly.

szeli1 commented:
I have moved the `ThreadedExportManager` to `Engine.h` because ideally one instance should handle all the exporting.
I also reworked `ThreadedExportManager.h` and removed unused code.

regulus79 commented:
Testing this on Arch Linux + GNOME

I tried opening a .wav sample and rightclicking and exporting, but even after entering a filename and pressing Save, it doesn't create a file. I tried multiple times with different filenames and different samples, but it didn't work.

For some reason when you click the "Export sample buffer" button, it opens the default gnome file explorer thing, not the lmms file dialog. But when double clicking on a sample to open a file, it opens the lmms file dialog.

![image](https://github.com/user-attachments/assets/22afa96e-3552-473c-81ca-8fa399fdba75)

szeli1 commented:
> I tried opening a .wav sample and rightclicking and exporting, but even after entering a filename and pressing Save, it doesn't create a file. I tried multiple times with different filenames and different samples, but it didn't work.

Thanks for noticing this, the issue was an if statement that checked when to abort. I have fixed it.

> For some reason when you click the "Export sample buffer" button, it opens the default gnome file explorer thing, not the lmms file dialog. But when double clicking on a sample to open a file, it opens the lmms file dialog.

It was decided to use the default file browser to export.

Thanks for your testing!

szeli1 commented on include/FlacExporter.h:
```diff
@@ -0,0 +1,53 @@
+	bool getIsSuccesful() const;
+	
+private:
+	bool m_isSuccesful;
```
I would keep the code this way, the same kind of code is used in `AudioFileFlac` and I think throw handling code will complicate things needlessly.

AW1534 commented:
> It was decided to use the default file browser to export.
Why?

szeli1 commented:
> Could you link the discussion or explain why?

Usually custom file browsers are subpar compared to the file browsers provided by the os. This is the case with QT. Users can change what kind of file browser they use (on linux), but they can't replace QT's file browser.

AW1534 commented:
> > Could you link the discussion or explain why?
> 
> Usually custom file browsers are subpar compared to the file browsers provided by the os. This is the case with QT. Users can change what kind of file browser they use (on linux), but they can't replace QT's file browser.

I agree with that, but switching to the native file browser is not as simple as it seems (as @regulus79, @tresf and I discovered in #7856). So if it's not strictly necessary for this PR, you should stick with the Qt browser for now, and #7856 will change it when it's ready.

JohannesLorenz commented on src/core/ThreadedExportManager.cpp:
```diff
@@ -0,0 +1,101 @@
+
+ThreadedExportManager::~ThreadedExportManager()
+{
+	stopExporting();
```
Let us assume I close LMMS while there are still buffers left to be rendered... Then this here will set `m_isThreadRunning` to `false`. Hence, `threadedExportMethod` will see `shouldRun` as `false` and will discard all pending requests - which means data loss. So, why not remove the `m_isThreadRunning` and just let `stopExporting` call `m_thread->join()`, which will wait until all pending requests are finished?

szeli1 replied:
I would expect things to be lost when I start to export something and then close the application. However nothing indicates that the exporting was finished.

Maybe use the callback function?
